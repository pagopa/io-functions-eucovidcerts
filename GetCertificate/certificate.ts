/* eslint-disable sort-keys */
/**
 * Certificate parsing utilities
 */
import { NonEmptyString } from "@pagopa/ts-commons/lib/strings";
import { Either, left } from "fp-ts/lib/Either";
import * as t from "io-ts";
import { WithinRangeInteger } from "@pagopa/ts-commons/lib/numbers";
import { DateFromString } from "@pagopa/ts-commons/lib/dates";
import { toTWithMap, toTWithMapOptional } from "../utils/conversions";
import { diseaseAgentTargeted } from "./valuesets/diseaseAgentTargeted";
import { vaccineProphylaxis } from "./valuesets/vaccineProphylaxis";
import { vaccineMedicinalProduct } from "./valuesets/vaccineMedicinalProduct";
import { marketingAuthorizationHolder } from "./valuesets/marketingAuthorizationHolders";
import { labTestTypes } from "./valuesets/labTestTypes";
import { labTestManufactorers } from "./valuesets/labTestManufactorers";
import { testResults } from "./valuesets/testResults";

type PersonName = t.TypeOf<typeof PersonName>;
const PersonName = t.interface({
  fn: t.string, // Familiy Name
  fnt: NonEmptyString, // Standardized Family Name
  gn: t.string, // Given Name
  gnt: t.string // Standardized Given Name
});

export type VaccinationEntry = t.TypeOf<typeof VaccinationEntry>;
export const VaccinationEntry = t.interface({
  tg: t.string.pipe(toTWithMap(diseaseAgentTargeted)), // disease or agent targeted
  vp: t.string.pipe(toTWithMap(vaccineProphylaxis)), // vaccine or prophylaxis
  mp: t.string.pipe(toTWithMap(vaccineMedicinalProduct)), // vaccine medicinal product
  ma: t.string.pipe(toTWithMap(marketingAuthorizationHolder)), // Marketing Authorization Holder
  dn: WithinRangeInteger(1, 9), // Dose Number
  sd: WithinRangeInteger(1, 9), // Total Series of Doses
  dt: DateFromString, // Date of Vaccination
  co: NonEmptyString, // Country of Vaccination
  is: NonEmptyString, // Certificate Issuer
  ci: NonEmptyString // Unique Certificate Identifier: UVCI
});

export type TestEntry = t.TypeOf<typeof TestEntry>;
export const TestEntry = t.intersection([
  t.interface({
    tg: t.string.pipe(toTWithMap(diseaseAgentTargeted)), // disease or agent targeted
    tt: t.string.pipe(toTWithMap(labTestTypes)), // covid-19 Lab Test Types
    sc: DateFromString, // Date/Time of Sample Collection
    tr: t.string.pipe(toTWithMap(testResults)), // Test Result
    tc: t.string, // Testing Centre
    co: t.string, // Country of Test
    is: t.string, // Issuer
    ci: t.string // Unique Certificate Identifier: UVCI
  }),
  t.partial({
    nm: t.string, // test name
    dr: DateFromString, // Date/Time of Test Result
    ma: t.string.pipe(toTWithMapOptional(labTestManufactorers)) // lab test manufactorers
  })
]);

export type RecoveryEntry = t.TypeOf<typeof RecoveryEntry>;
export const RecoveryEntry = t.intersection([
  t.interface({
    tg: t.string.pipe(toTWithMap(diseaseAgentTargeted)), // disease or agent targeted
    fr: DateFromString, // ISO 8601 Date of First Positive Test Result
    co: t.string, // Country of Test
    is: t.string, // Certificate Issuer
    df: DateFromString, // ISO 8601 Date: Certificate Valid From
    du: DateFromString, // ISO 8601 Date: Certificate Valid Until
    ci: t.string // Unique Certificate Identifier: UVCI
  }),
  t.partial({})
]);

export const VacCertificate = t.interface({
  ver: NonEmptyString,
  nam: PersonName,
  dob: DateFromString,
  v: t.readonlyArray<typeof VaccinationEntry>(VaccinationEntry)
});
export const TestCertificate = t.interface({
  ver: NonEmptyString,
  nam: PersonName,
  dob: DateFromString,
  t: t.readonlyArray<typeof TestEntry>(TestEntry)
});
export const RecoveryCertificate = t.interface({
  ver: NonEmptyString,
  nam: PersonName,
  dob: DateFromString,
  r: t.readonlyArray<typeof RecoveryEntry>(RecoveryEntry)
});

export type Certificates = t.TypeOf<typeof Certificates>;
export const Certificates = t.union([
  VacCertificate,
  TestCertificate,
  RecoveryCertificate
]);

/**
 * Parse a given base64 string representing a qrcode image to extract Certificate's informations
 *
 * @param qrcode
 * @returns either the Certificate object or a parsing error message
 */
export const parseQRCode = (_qrcode: string): Either<string, Certificates> =>
  left(`QRCode parsing not yet implemented`);
