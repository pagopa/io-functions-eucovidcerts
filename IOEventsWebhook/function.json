{
  "bindings": [
    {
      "authLevel": "function",
      "type": "httpTrigger",
      "direction": "in",
      "name": "req",
      "route": "api/v1/io-events-webhook",
      "methods": [
        "post"
      ]
    },
    {
      "type": "queue",
      "direction": "out",
      "name": "outputFiscalCode",
      "queueName": "%EUCOVIDCERT_NOTIFY_NEW_PROFILE_QUEUE_NAME%",
      "connection": "QueueStorageConnection"
    },
    {
      "type": "http",
      "direction": "out",
      "name": "$return"
    }
  ],
  "scriptFile": "../dist/IOEventsWebhook/index.js"
}