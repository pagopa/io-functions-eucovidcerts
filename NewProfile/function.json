{
  "bindings": [
    {
      "queueName": "%EUCOVIDCERT_NOTIFY_QUEUE_NAME%",
      "connection": "AppQueueStorageConnection",
      "name": "inputFiscalCode",
      "type": "queueTrigger",
      "direction": "in"
    },
    {
      "type": "queue",
      "direction": "out",
      "name": "outputFiscalCode",
      "queueName": "notify-new-profile",
      "connection": "QueueStorageConnection"
    }
  ],
  "scriptFile": "../dist/NewProfile/index.js"
}
