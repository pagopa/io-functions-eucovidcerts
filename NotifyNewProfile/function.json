{
  "bindings": [
    {
      "queueName": "notify-new-profile",
      "connection": "QueueStorageConnection",
      "name": "fiscalCode",
      "type": "queueTrigger",
      "direction": "in"
    }
  ],
  "scriptFile": "../dist/NotifyNewProfile/index.js"
}
