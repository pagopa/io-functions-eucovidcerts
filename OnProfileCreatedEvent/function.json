{
  "bindings": [
    {
      "queueName": "%EUCOVIDCERT_PROFILE_CREATED_QUEUE_NAME%",
      "connection": "EventsQueueStorageConnection",
      "name": "inputFiscalCode",
      "type": "queueTrigger",
      "direction": "in"
    },
    {
      "type": "queue",
      "direction": "out",
      "name": "outputFiscalCode",
      "queueName": "%EUCOVIDCERT_NOTIFY_NEW_PROFILE_QUEUE_NAME%",
      "connection": "QueueStorageConnection"
    }
  ],
  "scriptFile": "../dist/OnProfileCreatedEvent/index.js"
}