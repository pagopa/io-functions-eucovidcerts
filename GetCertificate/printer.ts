/* eslint-disable sort-keys */
import {
  PreferredLanguage,
  PreferredLanguageEnum
} from "@pagopa/io-functions-commons/dist/generated/definitions/PreferredLanguage";
import * as o from "fp-ts/lib/Option";
import * as moment from "moment-timezone";
import { match } from "ts-pattern";
import { pipe } from "fp-ts/lib/function";
import { SupportedLanguage } from "../utils/conversions";
import {
  Certificates,
  TestCertificate,
  VacCertificate,
  VaccinationEntry,
  TestEntry,
  RecoveryCertificate,
  RecoveryEntry
} from "./certificate";
import * as vacDetailsEn from "./markdown/eucovidcertDetailsVaccinationEn";
import * as vacDetailsIt from "./markdown/eucovidcertDetailsVaccinationIt";
import * as vacDetailsDe from "./markdown/eucovidcertDetailsVaccinationDe";
import * as testDetailsIt from "./markdown/eucovidcertDetailsTestIt";
import * as testDetailsEn from "./markdown/eucovidcertDetailsTestEn";
import * as testDetailsDe from "./markdown/eucovidcertDetailsTestDe";
import * as recoveryDetailsIt from "./markdown/eucovidcertDetailsRecoveryIt";
import * as recoveryDetailsEn from "./markdown/eucovidcertDetailsRecoveryEn";
import * as recoveryDetailsDe from "./markdown/eucovidcertDetailsRecoveryDe";
import * as vacInfoMultilanguage from "./markdown/eucovidcertInfoVaccinationEn";
import * as vacInfoGerman from "./markdown/eucovidcertInfoVaccinationDe";
import * as expiredInfoIt from "./markdown/eucovidcertExpiredInfoIt";
import * as expiredInfoEn from "./markdown/eucovidcertExpiredInfoEn";
import * as expiredInfoDe from "./markdown/eucovidcertExpiredInfoDe";

import { labTestTypes, molecularTest } from "./valuesets/labTestTypes";

const TIME_ZONE = "Europe/Rome";

const DATE_FORMAT_EN = "YYYY-MM-DD";
const DATE_FORMAT_ITA = "DD-MM-YYYY";

const dateFormatForLanguage: { [key in SupportedLanguage]: string } = {
  [PreferredLanguageEnum.it_IT]: DATE_FORMAT_ITA,
  [PreferredLanguageEnum.en_GB]: DATE_FORMAT_EN,
  [PreferredLanguageEnum.de_DE]: DATE_FORMAT_EN
};

const DATE_TIME_FORMAT_EN = "YYYY-MM-DD HH:mm";
const DATE_TIME_FORMAT_ITA = "DD-MM-YYYY HH:mm";

const dateAndTimeFormatForLanguage: { [key in SupportedLanguage]: string } = {
  [PreferredLanguageEnum.it_IT]: DATE_TIME_FORMAT_ITA,
  [PreferredLanguageEnum.en_GB]: DATE_TIME_FORMAT_EN,
  [PreferredLanguageEnum.de_DE]: DATE_TIME_FORMAT_EN
};
interface IPrintersForLanguage {
  readonly expiredInfoPrinter: () => string;
  readonly detailVaccinePrinter: (v: VaccinationEntry) => string;
  readonly detailTestPrinter: (t: TestEntry) => string;
  readonly detailRecoveryPrinter: (r: RecoveryEntry) => string;
  readonly infoVaccinePrinter: (c: Certificates) => string;
  readonly infoTestPrinter: (c: Certificates) => string;
  readonly infoRecoveryPrinter: (c: Certificates) => string;
}

const printersConfigurations: {
  [key in SupportedLanguage]: IPrintersForLanguage;
} = {
  [PreferredLanguageEnum.it_IT]: {
    expiredInfoPrinter: expiredInfoIt.getInfoPrinter,
    detailVaccinePrinter: vacDetailsIt.getDetailPrinter,
    detailTestPrinter: testDetailsIt.getDetailPrinter,
    detailRecoveryPrinter: recoveryDetailsIt.getDetailPrinter,
    infoVaccinePrinter: vacInfoMultilanguage.getInfoPrinter,
    infoTestPrinter: vacInfoMultilanguage.getInfoPrinter,
    infoRecoveryPrinter: vacInfoMultilanguage.getInfoPrinter
  },
  [PreferredLanguageEnum.en_GB]: {
    expiredInfoPrinter: expiredInfoEn.getInfoPrinter,
    detailVaccinePrinter: vacDetailsEn.getDetailPrinter,
    detailTestPrinter: testDetailsEn.getDetailPrinter,
    detailRecoveryPrinter: recoveryDetailsEn.getDetailPrinter,
    infoVaccinePrinter: vacInfoMultilanguage.getInfoPrinter,
    infoTestPrinter: vacInfoMultilanguage.getInfoPrinter,
    infoRecoveryPrinter: vacInfoMultilanguage.getInfoPrinter
  },
  [PreferredLanguageEnum.de_DE]: {
    expiredInfoPrinter: expiredInfoDe.getInfoPrinter,
    detailVaccinePrinter: vacDetailsDe.getDetailPrinter,
    detailTestPrinter: testDetailsDe.getDetailPrinter,
    detailRecoveryPrinter: recoveryDetailsDe.getDetailPrinter,
    infoVaccinePrinter: vacInfoGerman.getInfoPrinter,
    infoTestPrinter: vacInfoGerman.getInfoPrinter,
    infoRecoveryPrinter: vacInfoGerman.getInfoPrinter
  }
};

export const defaultPrinter: IPrintersForLanguage =
  printersConfigurations[PreferredLanguageEnum.en_GB];

/**
 * Return the correct IPrintersForLanguage based on preferred language,
 * or default IPrintersForLanguage if no language is selected
 *
 * @param lang the preferred language, if any
 * @returns
 */
export const getPrinterForLanguage = (
  lang: o.Option<PreferredLanguage>
): IPrintersForLanguage =>
  pipe(
    lang,
    o.chain(l =>
      o.fromNullable(
        (printersConfigurations as Record<
          PreferredLanguage,
          IPrintersForLanguage
        >)[l]
      )
    ),
    o.getOrElse(() => defaultPrinter)
  );

/**
 * Returns the Detail markdown filled with data from the input Certificate
 * based on preferred language and the Certificate type (vac, test, rec)
 *
 * @param lang the preferred language
 * @param c the Certificate
 * @returns a string containing the Info markdown
 */
export const printDetails = (
  lang: o.Option<PreferredLanguage>,
  c: Certificates
): string =>
  match(c)
    .when(VacCertificate.is, vc =>
      getPrinterForLanguage(lang).detailVaccinePrinter(vc.v[0])
    )
    .when(TestCertificate.is, tc =>
      getPrinterForLanguage(lang).detailTestPrinter(tc.t[0])
    )
    .when(RecoveryCertificate.is, rc =>
      getPrinterForLanguage(lang).detailRecoveryPrinter(rc.r[0])
    )
    .exhaustive();

/**
 * Returns the Info markdown filled with data from the input Certificate
 * based on preferred language
 *
 * @param lang the preferred language
 * @param c the Certificate
 * @returns a string containing the Info markdown
 */
export const printInfo = (
  lang: o.Option<PreferredLanguage>,
  c: Certificates
): string => getPrinterForLanguage(lang).infoVaccinePrinter(c);

/**
 * Returns UVCI of the first certificate (either Vaccination, Test or Recovery)
 *
 * @param _lang the preferred language (unused)
 * @param c the Certificate
 * @returns the UVCI
 */
export const printUvci = (
  _lang: o.Option<PreferredLanguage>,
  c: Certificates
): string =>
  match(c)
    .when(VacCertificate.is, cv => cv.v[0].ci)
    .when(TestCertificate.is, ct => ct.t[0].ci)
    .when(RecoveryCertificate.is, cr => cr.r[0].ci)
    .exhaustive();

/**
 * Returns the markdown info text to be rendered when a Certificate is found to be expired
 * based on preferred language
 *
 * @param lang the preferred language
 * @returns a string containing the Info markdown
 */
export const printExpiredInfo = (lang: o.Option<PreferredLanguage>): string =>
  getPrinterForLanguage(lang).expiredInfoPrinter();

/**
 * Check if Vaccination has ended
 *
 * @param v the VaccinationEntry to check
 * @returns true if process ended, false otherwise
 */
export const isVaccinationProcessEnded = (v: VaccinationEntry): boolean =>
  v.dn >= v.sd; // we may have more doses than the total in case of additional, "booster doses"

/**
 * Test validity in hours, based on type
 *
 * @param v the VaccinationEntry to check
 * @returns the validity of the test, in hours
 */
export const testValidity = (v: TestEntry): string =>
  v.tt === labTestTypes[molecularTest] ? "72" : "48";

/**
 * Format date value based on language
 *
 * @param d the date
 * @param _lang the preferred language
 * @returns a formatted date
 */
export const formatDate = (d: Date, lang: SupportedLanguage): string =>
  pipe(dateFormatForLanguage[lang], format =>
    moment(d)
      .tz(TIME_ZONE)
      .format(format)
  );

/**
 * Format date and time value based on language
 *
 * @param d the date
 * @param _lang the preferred language
 * @returns a formatted date with time
 */
export const formatDateAndTime = (d: Date, lang: SupportedLanguage): string =>
  // eslint-disable-next-line sonarjs/no-identical-functions
  pipe(dateAndTimeFormatForLanguage[lang], format =>
    moment(d)
      .tz(TIME_ZONE)
      .format(format)
  );

/**
 * The issuer (is) field value used by Italy Healthcare Department
 */
export const ITALY_HEALTHCARE_ISSUER = "IT";

export const HEALTHCARE_DEP_IT = "Ministero della Salute";
export const HEALTHCARE_DEP_EN = "Ministry of Health";
export const HEALTHCARE_DEP_DE = "Gesundheitsministerium";

const healthcareIssuerForLanguage: { [key in SupportedLanguage]: string } = {
  [PreferredLanguageEnum.it_IT]: HEALTHCARE_DEP_IT,
  [PreferredLanguageEnum.en_GB]: HEALTHCARE_DEP_EN,
  [PreferredLanguageEnum.de_DE]: HEALTHCARE_DEP_DE
};

/**
 * Format Certificate Issuer based on its value and preferred language
 *
 * @param c the issuer certificate value
 * @param _lang the preferred language
 * @returns a formatted Certificate Issuer string
 */
export const formatCertificateIssuer = (
  c: string,
  lang: SupportedLanguage
): string =>
  c !== ITALY_HEALTHCARE_ISSUER ? c : healthcareIssuerForLanguage[lang];
