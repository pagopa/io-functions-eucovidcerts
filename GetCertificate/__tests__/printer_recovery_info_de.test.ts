import { PreferredLanguageEnum } from "@pagopa/io-functions-commons/dist/generated/definitions/PreferredLanguage";
import { getOrElseW } from "fp-ts/lib/Either";
import { pipe } from "fp-ts/lib/function";
import { some } from "fp-ts/lib/Option";

import { Certificates } from "../certificate";
import { printInfo } from "../printer";

const language = PreferredLanguageEnum.de_DE;

describe(`Printer - Recovery - Info - ${language}`, () => {
  it("should print its markdown", () => {
    const certificate = pipe(
      Certificates.decode({
        ver: "1.0.0",
        nam: {
          fn: "Di Caprio",
          fnt: "DI<CAPRIO",
          gn: "MarilÃ¹ Teresa",
          gnt: "MARILU<TERESA"
        },
        dob: "1977-06-16",
        r: [
          {
            tg: "840539006",
            fr: "2021-04-10",
            co: "IT",
            is: "IT",
            df: "2021-04-20",
            du: "2021-09-10",
            ci: "01ITE7300E1AB2A84C719004F103DCB1F70A#6"
          }
        ]
      }),
      getOrElseW(_ => {
        throw "Error decoding object";
      })
    );

    const result = printInfo(some(language), certificate);
    expect(result).toMatchSnapshot();
  });
});
