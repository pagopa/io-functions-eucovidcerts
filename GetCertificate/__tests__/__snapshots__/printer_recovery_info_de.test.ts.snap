// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Printer - Recovery - Info - de_DE should print its markdown 1`] = `
"
Surname(s) and forename(s)  
*Nachname und Vorname*  
**Di Caprio Maril√π Teresa**  

Date of birth  
*Geburtsdatum (JJJJ-MM-TT)*  
**1977-06-16**  

Unique Certificate Identifier  
*Eindeutige Kennung des Zertifikats*  
**01ITE7300E1AB2A84C719004F103DCB1F70A#6**  

[Kopiere die Kennung](iohandledlink://copy:01ITE7300E1AB2A84C719004F103DCB1F70A#6)
<br/><br/><br/><br/><br/>
"
`;
